shotover/src/codec/cassandra.rs

```diff
@@ -67,96 +153,246 @@ impl CassandraDecoder {
+
+                    let header_crc24 = ((header >> 24) & 0xffffff) as i32;
+                    let computed_crc = cassandra_protocol::crc::crc24(&header.to_le_bytes()[..3]);
+
+                    if header_crc24 != computed_crc {
+                        return Err(anyhow!(format!(
+                            "Header CRC mismatch - expected {header_crc24}, found {computed_crc}."
+                        )));
+                    }
+
+                    let payload_length = (header & 0x1ffff) as usize;
+                    let payload_end = UNCOMPRESSED_FRAME_HEADER_LENGTH + payload_length;
+
+                    let frame_end = payload_end + FRAME_TRAILER_LENGTH;
+
+                    let payload_crc32 =
+                        u32::from_le_bytes(frame_bytes[payload_end..frame_end].try_into().unwrap());
+
+                    let computed_crc = cassandra_protocol::crc::crc32(
+                        &frame_bytes[UNCOMPRESSED_FRAME_HEADER_LENGTH..payload_end],
+                    );
+                    if payload_crc32 != computed_crc {
+                        return Err(anyhow!(format!(
+                            "Payload CRC mismatch - read {payload_crc32}, computed {computed_crc}."
+                        )));
+                    }
+
+                    let self_contained = (header & (1 << 17)) != 0;
+                    if !self_contained {
+                        unimplemented!("Cannot support non-self contained frames yet");
+                    }
+
+                    frame_bytes.advance(UNCOMPRESSED_FRAME_HEADER_LENGTH);
+                    let mut payload = frame_bytes.split_to(payload_length);
+
+                    let mut envelopes: Vec<Message> = vec![];
+
+                    while !payload.is_empty() {
+                        let body_len =
+                            i32::from_be_bytes(payload[5..9].try_into().unwrap()) as usize;
+
+                        let envelope_len = ENVELOPE_HEADER_LEN + body_len;
+
+                        if envelope_len > payload.len() {
+                            return Err(anyhow!(format!(
+                                "envelope length {} is longer than payload length {}",
+                                envelope_len,
+                                payload.len()
+                            ),));
+                        }
+
+                        let envelope = payload.split_to(envelope_len);
+
+                        envelopes.push(Message::from_bytes(
+                            envelope.freeze(),
+                            crate::message::ProtocolType::Cassandra {
+                                compression: Compression::None,
+                            },
+                        ));
+                    }
+
+                    Ok(envelopes)
+                }
+                _ => unimplemented!(),
+            },
+            (_, _) => {
+                let bytes = src.split_to(frame_len);
+                tracing::debug!(
```

Comment:
Can we put this in a `for message in messages {` after the loop so that it still runs for v5?